extmod/machine_usb_device.c

```diff
@@ -0,0 +1,343 @@
+    { MP_ROM_QSTR(MP_QSTR_ep_max), MP_OBJ_NEW_SMALL_INT(0) },
+    { MP_ROM_QSTR(MP_QSTR_str_max), MP_OBJ_NEW_SMALL_INT(1) },
+    { MP_ROM_QSTR(MP_QSTR_desc_dev), MP_ROM_PTR(&builtin_default_desc_dev_obj)  },
+    { MP_ROM_QSTR(MP_QSTR_desc_cfg), mp_const_empty_bytes },
+};
+static MP_DEFINE_CONST_DICT(usb_device_builtin_none_dict, usb_device_builtin_none_dict_table);
+
+MP_DEFINE_CONST_OBJ_TYPE(
+    mp_type_usb_device_builtin_none,
+    MP_QSTR_BUILTIN_NONE,
+    MP_TYPE_FLAG_NONE,
+    locals_dict, &usb_device_builtin_none_dict
+    );
+
+static const mp_rom_map_elem_t usb_device_locals_dict_table[] = {
+    { MP_ROM_QSTR(MP_QSTR_config), MP_ROM_PTR(&usb_device_config_obj) },
+    { MP_ROM_QSTR(MP_QSTR_submit_xfer), MP_ROM_PTR(&usb_device_submit_xfer_obj) },
+    { MP_ROM_QSTR(MP_QSTR_active), MP_ROM_PTR(&usb_device_active_obj) },
+    { MP_ROM_QSTR(MP_QSTR_stall), MP_ROM_PTR(&usb_device_stall_obj) },
+
+    // Built-in driver constants
+    { MP_ROM_QSTR(MP_QSTR_BUILTIN_NONE), MP_ROM_PTR(&mp_type_usb_device_builtin_none) },
+
+    #if !HAS_BUILTIN_DRIVERS
+    // No builtin-in drivers, so BUILTIN_DEFAULT is BUILTIN_NONE
+    { MP_ROM_QSTR(MP_QSTR_BUILTIN_DEFAULT), MP_ROM_PTR(&mp_type_usb_device_builtin_none) },
+    #else
+    { MP_ROM_QSTR(MP_QSTR_BUILTIN_DEFAULT), MP_ROM_PTR(&mp_type_usb_device_builtin_default) },
+
+    // Specific driver constant names are to support future switching of built-in drivers,
+    // but currently only one is present and it maps directly to BUILTIN_DEFAULT
+    #if MICROPY_HW_USB_CDC && !MICROPY_HW_USB_MSC
+    { MP_ROM_QSTR(MP_QSTR_BUILTIN_CDC), MP_ROM_PTR(&mp_type_usb_device_builtin_default) },
+    #endif
+    #if MICROPY_HW_USB_MSC && !MICROPY_HW_USB_CDC
+    { MP_ROM_QSTR(MP_QSTR_BUILTIN_MSC), MP_ROM_PTR(&mp_type_usb_device_builtin_default) },
+    #endif
+    #if MICROPY_HW_USB_CDC && MICROPY_HW_USB_MSC
+    { MP_ROM_QSTR(MP_QSTR_BUILTIN_CDC_MSC), MP_ROM_PTR(&mp_type_usb_device_builtin_default) },
+    #endif
+    #endif // !HAS_BUILTIN_DRIVERS
+};
+static MP_DEFINE_CONST_DICT(usb_device_locals_dict, usb_device_locals_dict_table);
+
+static void usb_device_attr(mp_obj_t self_in, qstr attr, mp_obj_t *dest) {
+    mp_obj_usb_device_t *self = MP_OBJ_TO_PTR(self_in);
+    if (dest[0] == MP_OBJ_NULL) {
+        // Load attribute.
+        if (attr == MP_QSTR_builtin_driver) {
+            dest[0] = self->builtin_driver;
+        } else {
+            // Continue lookup in locals_dict.
+            dest[1] = MP_OBJ_SENTINEL;
+        }
+    } else if (dest[1] != MP_OBJ_NULL) {
+        // Store attribute.
+        if (attr == MP_QSTR_builtin_driver) {
+            if (self->active) {
+                mp_raise_OSError(MP_EINVAL); // Need to deactivate first
+            }
+
+            if (
+                dest[1] == MP_OBJ_TO_PTR(&mp_type_usb_device_builtin_none)
+                #if HAS_BUILTIN_DRIVERS
+                || dest[1] == MP_OBJ_TO_PTR(&mp_type_usb_device_builtin_default)
```

Comment:
You don't really need this check for a valid builtin.  The code in this PR will still work perfectly fine if you store any arbitrary object in `self->builtin_driver`.

I can see why you'd want some error checking, but is it needed?

Reply:
The user won't get the behaviour they expect if they assign a different object here, the other pieces of code only check for "none" or "default" for built-in drivers. It doesn't look at any of the fields on the object from the C side.

That's something that maybe could be changed in the future by encoding more information about the drivers inside the object itself, so an arbitrary object would work as expected, but currently the logic that reads this field is comparing the address.

Or do you mean it's better to accept anything here and perhaps behave unexpectedly, rather than error out?

Reply:
> Or do you mean it's better to accept anything here and perhaps behave unexpectedly, rather than error out?

Yes.  I'm just thinking from a small/minimal code size point of view.  This driver is low-level so doesn't need to be 100% user friendly.

Reply:
Got it, have changed